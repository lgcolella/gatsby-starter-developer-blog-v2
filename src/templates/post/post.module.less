@import (reference) '../../style/index.less';

.layout {
    width: 100%;
    @media @tablet, @desktop {
        margin-top: 0px !important;
    }
}

.content {
    box-sizing: border-box;
    margin: auto;
    padding: @padding-medium;
    width: 80%;
    @media @mobile {
        width: 100%;
    }
}
