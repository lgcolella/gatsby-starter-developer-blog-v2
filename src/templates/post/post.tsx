/* Vendor imports */
import React from 'react'
import { graphql } from 'gatsby'
/* App imports */
import {
  AvailableTag,
  ChildImageSharpFluid,
  AvailableLanguage,
} from '../../types'
import Layout from '../../components/layout'
import SEO from '../../components/seo'
import Heading from './heading'
import ArticleHeading from './article-heading'
import Article from './article'
import Share from './share'
import SuggestedPosts from './suggested-posts'
import Config from '../../../config'
import Utils from '../../utils'
import Comments from './comments'
const style = require('./post.module.less')

const Post = ({ data, pageContext }: Props) => {
  const { id, html, frontmatter, timeToRead } = data.markdownRemark
  const { title, date, tags, cover, path, excerpt } = frontmatter
  const translations =
    pageContext.translations.length > 1 ? pageContext.translations : undefined
  const img = cover.childImageSharp.fluid
  const canonicalUrl = Utils.resolveUrl(Config.siteUrl, Config.pathPrefix, path)
  const suggestedPosts = Utils.getSuggestedPosts(
    data.markdownRemark,
    data.allMarkdownRemark as any,
    3
  )

  return (
    <Layout className={style.layout}>
      <SEO
        title={title}
        description={excerpt}
        path={path}
        contentType='article'
        imageUrl={img.src}
        keywords={tags}
        translations={translations}
      />
      <Heading title={title} tags={tags} cover={img} coverTitle={excerpt} />
      <div className={style.content}>
        <ArticleHeading
          excerpt={excerpt}
          date={date}
          time={timeToRead}
          translations={translations}
        />
        <Article html={html} />
        <Share pageCanonicalUrl={canonicalUrl} title={title} />
        <Comments url={canonicalUrl} id={id} title={title} />
        <SuggestedPosts posts={suggestedPosts} />
      </div>
    </Layout>
  )
}

export interface Props {
  data: {
    markdownRemark: {
      id: string
      html: string
      timeToRead: number
      fileAbsolutePath: string
      frontmatter: {
        title: string
        date: string
        tags: AvailableTag[]
        path: string
        excerpt: string
        cover: {
          childImageSharp: ChildImageSharpFluid
        }
      }
    }
    allMarkdownRemark: {
      edges: Array<{
        node: {
          fileAbsolutePath: string
          frontmatter: {
            path: string
            title: string
            tags: AvailableTag[]
            excerpt: string
            cover: {
              childImageSharp: ChildImageSharpFluid
            }
          }
        }
      }>
    }
  }
  pageContext: {
    translations: Array<{
      hreflang: AvailableLanguage
      path: string
    }>
  }
}

export const pageQuery = graphql`
  query($postId: String!) {
    markdownRemark(id: { eq: $postId }) {
      id
      html
      timeToRead
      fileAbsolutePath
      frontmatter {
        title
        date(formatString: "YYYY-MM-DD")
        tags
        path
        excerpt
        cover {
          childImageSharp {
            fluid(maxWidth: 1000) {
              ...GatsbyImageSharpFluid_noBase64
            }
          }
        }
      }
    }
    allMarkdownRemark {
      edges {
        node {
          fileAbsolutePath
          frontmatter {
            path
            title
            tags
            excerpt
            cover {
              childImageSharp {
                fluid(maxWidth: 600) {
                  ...GatsbyImageSharpFluid_tracedSVG
                }
              }
            }
          }
        }
      }
    }
  }
`
export default Post
