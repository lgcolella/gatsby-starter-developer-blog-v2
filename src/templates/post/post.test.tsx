/* Vendor imports */
import React from 'react'
import { shallow } from 'enzyme'
/* App imports */
import Post from './index'
import { Props } from './post'
import { ImageSharpFluid } from '../../../tests/mocks'

describe('Post template', () => {
  const data: Props['data'] = {
    markdownRemark: {
      id: '65f08b56-df9b-5392-9dcf-894ea64c82be',
      html: '<div><h2>Sub title</h2><p>Article paragraph</p></div>',
      timeToRead: 6,
      fileAbsolutePath: 'C:/Documents/absolute-path-to-the-main-post',
      frontmatter: {
        title: 'main-post',
        date: '2019-07-23',
        tags: ['angular', 'javascript'],
        path: 'url-to-the-main-post',
        excerpt: 'excerpt for the main post',
        cover: {
          childImageSharp: ImageSharpFluid,
        },
      },
    },
    allMarkdownRemark: {
      edges: [
        // Article #1
        {
          node: {
            fileAbsolutePath: 'C:/Documents/path-to-the-article-1',
            frontmatter: {
              title: 'Title of article #1',
              path: 'path-to-the-article-1',
              tags: ['angular', 'javascript', 'laravel'],
              excerpt: 'an excertp for the article #1',
              cover: {
                childImageSharp: ImageSharpFluid,
              },
            },
          },
        },
        // Article #2
        {
          node: {
            fileAbsolutePath: 'C:/Documents/path-to-the-article-2',
            frontmatter: {
              title: 'Title of article #2',
              path: 'path-to-the-article-2',
              tags: ['nodejs', 'angular'],
              excerpt: 'an excertp for the article #2',
              cover: {
                childImageSharp: ImageSharpFluid,
              },
            },
          },
        },
      ],
    },
  }

  test('matches snapshot', () => {
    let wrapper = shallow(
      <Post
        data={data}
        pageContext={{
          translations: [
            { hreflang: 'it', path: '/it/version-of-main-post' },
            { hreflang: 'en', path: '/en/version-of-main-post' },
          ],
        }}
      />
    )

    expect(wrapper).toMatchSnapshot()
  })
})
