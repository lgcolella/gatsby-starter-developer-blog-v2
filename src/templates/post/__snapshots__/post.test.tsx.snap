// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Post template matches snapshot 1`] = `
<Layout
  className="layout"
  showHeader={true}
  title=""
>
  <SEO
    contentType="article"
    description="excerpt for the main post"
    imageUrl="/static/da8d0fc0af7f0d310517048e303b3fcf/a5a41/preview.jpg"
    keywords={
      Array [
        "angular",
        "javascript",
      ]
    }
    path="url-to-the-main-post"
    title="main-post"
    translations={
      Array [
        Object {
          "hreflang": "it",
          "path": "/it/version-of-main-post",
        },
        Object {
          "hreflang": "en",
          "path": "/en/version-of-main-post",
        },
      ]
    }
  />
  <Heading
    cover={
      Object {
        "aspectRatio": 1.7,
        "sizes": "(max-width: 600px) 100vw, 600px",
        "src": "/static/da8d0fc0af7f0d310517048e303b3fcf/a5a41/preview.jpg",
        "srcSet": "/static/da8d0fc0af7f0d310517048e303b3fcf/a5a41/preview.jpg",
      }
    }
    coverTitle="excerpt for the main post"
    tags={
      Array [
        "angular",
        "javascript",
      ]
    }
    title="main-post"
  />
  <div
    className="content"
  >
    <ArticleHeading
      date="2019-07-23"
      excerpt="excerpt for the main post"
      time={6}
      translations={
        Array [
          Object {
            "hreflang": "it",
            "path": "/it/version-of-main-post",
          },
          Object {
            "hreflang": "en",
            "path": "/en/version-of-main-post",
          },
        ]
      }
    />
    <Article
      html="<div><h2>Sub title</h2><p>Article paragraph</p></div>"
    />
    <Share
      pageCanonicalUrl="https://lgcolella.github.io/gatsby-starter-developer-blog-v2/url-to-the-main-post/"
      title="main-post"
    />
    <SuggestedPosts
      posts={Array []}
    />
  </div>
</Layout>
`;
